### DISCUSSION TITLE
Support for GPU/CUDA

### DISCUSSION THREAD

**Discussion by haixuanTao** [sonos/tract#688] [Ideas]
I tried this package on the exemple Pytorch and it seems that there's currently no GPU/CUDA support.

Is it a feature that is going to be added in the future? 

Thanks in advance for this package :)

**Comment by kali** [comment]
Thanks for your interest

This is correct. There is no CUDA or GPU support, and it is not on the roadmap. We are not putting too much energy on the PC platform, tract main focus being arm32 and arm64 devices. If something happens on the GPU, it won't come from us.

**Comment by haixuanTao** [comment]
Make sense.

Best option for GPU ONNX on rust at the moment seems to be https://github.com/nbigaouette/onnxruntime-rs

I'll close the issue.

**Comment by dvc94ch** [comment]
That might make sense for training the nn on a gpu. Support for gpu in tract is probably more along the lines of doing inference on a mobile gpu. Might make sense to reopen the issue as it's not resolved and other people might be interested in the status of this.

**Comment by dsseng** [comment]
+1 for reopening. Yes, I beleive support for, e.g., Vulkan, would boost the library capabilities. I have found this project and liked it because Rust is definitely more maintainable than C. With hardware acceleration tract will be a really great tool for mobile as well (mobile devices have GL/CL/Vulkan to accelerate inference, and those are a lot more power-efficient). I would like to work on some HW acceleration approach if such is wanted.

One of ideas is to create a loadable C-ABI libraries for accel methods. It would require creating an API and binary-stable interface to C-style libraries (built with Rust, C/C++ or many other languages), which are discovered and loaded by core using `libloading`. They would need to have attributes such as `AVX2` (arch features necessary to load,  btw, SIMD could also help accelerating inference). API should include generic buffer management functions (create, free, pass to dev/to CPU, get location (shm, unified memory, dedicated memory, RDMA etc), import/export platform buffer) and operators' implementations.

I propose implementing either `wgpu` or direct Vulkan backend. Former is preferred, as it should be able to activate GPU nearly on anything: Windows, Linux (incl. Android) and macOS + MVK with Vulkan backend, Metal for macOS and iOS, fallback OpenGL on Linux (and Android) with older GPUs, DirectX fallback for Windows when Vulkan is not available, WebGPU with WebGL fallback with WebAssembly.

**Reply by haixuanTao** [reply]
Hey, for info, I have a small library that does Deep Learning on `wgpu` called `wonnx` https://github.com/webonnx/wonnx/ . You can measure the GPU boost from tract with the cli, check: https://github.com/webonnx/wonnx/discussions/65 . Although, the project is much less mature than Tract.

I think that buffer management and shader optimization is really important if you want to port wgpu-acceleration to tract.

**Reply by dsseng** [reply]
Looks great! Connecting existing code together with rewriting could definitely help combining tract's flexibility with wonnx's hardware acceleration. What do you think about an acceleration architecture I proposed (libloading a lib which implements functions)? While API is only getting stable we might need to consider some different approach as well.

**Reply by kali** [reply]
Well, there should be some synergy here... plugging wonnx at tract-core level would make sense. Wonnx would benefit from tract support for nnef (and a bit of tensorflow) and would have to deal with a reduced set of operators (as tract-core is smaller than onnx).

**Reply by haixuanTao** [reply]
So I feel like you would like to create a C-ABI to be able to connect different accelerator providers in a similar fashion to ONNX Providers. ( Correct me if I'm wrong).

I think this could be interesting, but as a developer, going through C-ABI is going to make developing and debugging more complex. I personally feel like libloading in Rust is not recommended when not necessary ( life C FFI bindings).

**Reply by haixuanTao** [reply]
> Well, there should be some synergy here... plugging wonnx at tract-core level would make sense. Wonnx would benefit from tract support for nnef (and a bit of tensorflow) and would have to deal with a reduced set of operators (as tract-core is smaller than onnx).

I think that this would be great, as it would reduce the load of wonnx maintenance on the IR level, which is kind of the current main effort.
Do @sh7dm you want to have a go at it? I can look it up also, if there is community interest.

**Reply by haixuanTao** [reply]
I don't mind trying to maintain the GPU/WebGPU part of tract if we do it, to not overload current Tract Maintainers with this piece of software which is pretty unstable.

**Reply by kali** [reply]
I agree we should try to consider alternatives to libloading before jumping into this. A possible way to avoid that (which does not handle multiple configuration automativally) is to have the application code 1/ load the model with tract-onnx (obtaining a tract-core model) 2/ then instead of doing `.into_optimized()` and so on, hand the TypedModel to a separate accelerated runtime of its choice.

This separated accelerated runtime would parse the TypedModel, use as much of tract-core infrastructure as makes sense (maybe the Tensor struct for interfacing, maybe not even that, maybe use some CPU implementation from tract-core if not everything can be done in accelerated fashion).

This is how a couple of teams with specific runtimes are already proceeding.

**Reply by kali** [reply]
Well, bottom line is, this mapping only consumes tract-core, and not the other way around, so it can be done as a separate crate, inside or outside of tract repo. If/When we get something semi-working we can decide where we want it to live permanently.

**Reply by dsseng** [reply]
Into_optimized must not be skipped I think, as far as I understood your docs. I definitely want to collaborate with both of you on this project, because it seems to absolutely fill the niche of inference frameworks (all ML that is still a place to conquer, training has already been splitted between TF and PyTorch). I'm not the greatest maintainer, but would like to do the coding part being interested in both GPUs and ML)

**Reply by kali** [reply]
into_optimized *must* be skipped. Most of the "optimisation" is model simplification and actually happens during the declutter stage. into_optimized performs cpu dependent optimisation, typically choosing which matrix multiplier kernel to use, among other thing that will make the network semantics very opaque and hard to translate to another runtime.

**Reply by haixuanTao** [reply]
Agreed. As GPU optimisation is completely different. That's great that you split the method into decluttering and optimization.

**Reply by kali** [reply]
Inspiration comes from LLVM really, and it does show in the wildly inconsistent naming. I think of tract-core as a MIR, while into_optimized translate into a LIR space. into_optimized actually calls the "codegen" method on operators.

**Reply by dsseng** [reply]
Ah, then I misread that. Exactly, if model simplification takes place earlier, we should split a new backend after there

**Reply by haixuanTao** [reply]
As your mentioning LLVM, have you checked MLIR ? And do you believe that it could be a better Intermediate representation in the future. I feel like ONNX and tensorflow-IR are both often not enough low-level. 

Just curious, as MLIR is still pretty young and support pretty low.

**Reply by dsseng** [reply]
Hmm, it should be definitely looked at, but for now it's too young, as you mentioned. Using LLVM and MLIR should be some long-term goals

**Comment by kali** [comment]
FYI, tract-linalg already provides runtime switchable implementations for some high cost operations (matrix multiplication, sigmoid and tanh, mostly). It is accelerated already on SIMD (with specific code for intel v2, and some armv7 and arm8 configurations).

One possible path to accelerating tract when a GPU is available is to implement the matrix multiplication on GPU. I think there is a MVP here with local changes only (in tract-linalg). We could then move on to lowering more operators in tract-linalg, discuss buffer locality and stuff, that would require some awareness from tract-core and tract-data.

One other option is to use tract-core as infrastructure and semantic pivot, but instead of optimizing the network in tract, substitute some operators with external ones from another crate that provides acceleration. This path is used by some teams who have specific hardware with capabilities they want to use without exposing themselves, as it allows things to happen outside of tract. This is not a requirement, such a crate could also be part of tract. Also, a source of friction for external consumer of tract-core is tract-core stability, on which I have not make commitments so far as I considered it internal, even if the API is de-facto pretty stable.

The reason why I closed this as an issue is not that I don't want it to happen: I don't have resources for it. tract is mostly a one-man-show so far, so I must pick my battles. If somebody else picks this one up and start working on it, I can help with guidance and discussing some changes to make it easier, but I can't do the heavy lifting. Until somebody does, this conversation is just daydreaming while I believe issues have to be actionable.

I'm gonna move this topic to the discussion section at this stage.

**Reply by dsseng** [reply]
Only accelerating linear algebra primitives might not improve or even introduce a separate problem: passing buffers all the way around. We might even constrain power by bus transfers this way.

```
CPU --bus-- GPU
Load      >
Next op
is matmul >
            matmul
          < copy
Got result,
addition
copy      >
            add
          < copy
```

We definitely want buffers to be stored in GPU memory and not transferred after each operation. Functions should be done fully in backend, while core holds pointers and instructs what to do with data referenced. Thus, operators should be done mostly in GPU. Linalg acceleration (as far as I understand) will require some stuff still being computed on CPU.

**Reply by kali** [reply]
I am aware moving data between CPU and GPU is expensive. The problem being, if we don't want to do it at all, we need *all* of tract-core operators mapped in the GPU. It's an all-of-nothing approach. If this is the path you'd like to follow, then using linalg as an abstraction layer is useless. Better use the second path, using tract to load and declutter models, and then translate from tract-core to whatever forms works for the GPU.

**Reply by dsseng** [reply]
Hmm, what does each crate do precisely? Sorry, I'm new to codebase since I discovered it yesterday.

**Reply by dsseng** [reply]
Acceleration architecture is like one of TensorFlow with PluggableDevice (lacks docs, but looks great).

**Reply by kali** [reply]
Let's start here: https://github.com/sonos/tract/blob/main/doc/intro.md :)

**Reply by kali** [reply]
About the buffer management cost... the rationale with "starting" with the matrix multiplier is that it is the most expensive thing to run, and the one GPU are the most suited for. So we may already have a gain by running these operations on GPU even factoring in the transfert cost, while leaving the cheaper stuff to run on the CPU. This would be a milestone / POC that do not require flipping all of tract codebase on its head.

**Reply by dsseng** [reply]
Thanks for docs, sorry for not finding it before. Will read and think more about structure and plugging GPU in there

**Reply by dsseng** [reply]
After discussing I can try to make a PoC with GPU and compare running time of examples, such as ALBERT (it is rather slow on my machine, 3.5 s). It's always better to bench rather than think about cost of such things.

**Comment by dsseng** [comment]
Trying to create a stateful extension for `Ops` within `linalg` crate to implement GPU ops in there first. I guess it would be zero-copy with integrated solutions like my Ryzen laptop or mobile/embedded SoC-s.

**Reply by dsseng** [reply]
But I do not believe that it is going to be zero-copy by default, might need some hacks around it.

**Comment by dsseng** [comment]
Are only 4x4, 4x1 and 3x2 matrices being multiplied? Weird. If larger ones are somehow splitted, how can that be disabled to handle full size on the GPU? The crate does not have enough internal docs to be understood and abstracted easily, I might need a direct consulting to get something done. Do you have some chat room like Matrix for that?

**Reply by kali** [reply]
Lol. No. tract can multiply matrices of any size. The strategy is pretty classic, looping over tiles computed by a micro kernels. See https://tech-blog.sonos.com/posts/the-anatomy-of-efficient-matrix-multipliers/ for an intro on the topic, or the various papers by GotoBLAS and BLIS teams.

You probably don't want to plug inside linalg to POC this. If you do so, you need to implement the MatMul trait and tweak the arch and  feature detection in linalg/src/lib.rs to plug yourself in. But it will be some work, because you will also need to implement the fusable operations. It may be better to plug also at the tract-core level, just like @haixuanTao is planning, but just substituing the MatMulUnary operator by a GPU based one.

**Reply by dsseng** [reply]
Hmm, will try. What about chat, would you be comfortable with that? It would be more productive because of ease to explore the internal structure.

**Reply by dsseng** [reply]
I ran into a lot of trouble while trying to plug it into linalg ops. E.g. ownership and Clone issues (you cannot easily clone shader object handles and so on to ensure safety for GPU stuff). So I need to explore more and, I guess, plug new impl instead of core or draw some extra split lines between and within modules (but that would be great, since we just upload an input buffer object and GPU does everything giving us a copyable output tensor in the end). Definitely would be quite a large PR.

**Reply by haixuanTao** [reply]
So, I'm ok to have a chat if need be, anytime you want. I am absolutely not familiar with Tract however. And so @Kali should participate if you have Tract questions.

From what I understood the plan is to use wonnx or part of the wonnx implementation for allocating the data and the computation but use Tract-Core instead of the current wonnx-parsing and optimisation code.

I feel like that is not what I think you have in mind. I won't be able to help you out in replacing linalg ops if that's what you are planning to do.

**Reply by dsseng** [reply]
Well, I don't feel the need in chat now, @ kali explained everything pretty well. I will see the code, but not sure about directly integrating Wonnx into Tract, maybe I will copy some parts of the code when contributing, but not everything directly (I guess architectures are different enough).

**Comment by dsseng** [comment]
Part of core features is platform-independent (Graph, Nodes, Inlets & Outlets etc). So, I guess, optim and ops should be split out into `backend-cpu`, and `backend-wgpu` crate should be created with analogic functions (then something like `backend-coral` can be made for specialized accelerators or other cases). They should be available as build flags and if both are selected, then `TRACT_BACKEND` env var will specify priority of backends.

**Reply by kali** [reply]
I think we will worry about that when we do have other backends... But the architecture I'm proposing is very extensible as alternative backends can be plugged at tract-core level, and are consumer only of tract core architecture so anybody can plug a full or partial backend at tract-core level. 

Also I think it is helpful to consider the CPU backend "special" as it will always work, and we need it during the declutter phase (for constant propagation for example).

**Reply by dsseng** [reply]
Yes, I fully agree. CPU backend cannot be disabled then? Hmm, well, it might be desirable if you have an embedded system where GPU is present, but you want something lightweight. Well, not a top priority now. What does plugging at core level mean precisely? What features of original core are left in place and which are made replaceable? Only optim and ops I guess should be done this way.

**Reply by dsseng** [reply]
And it looks like extensions bring some other ops as well. Am I right or are those mapped to core ops internally when loading model?

**Reply by kali** [reply]
On the tract-core TypedModel, obtained after decluttering, you can go over the model graph and map it entirely to GPU. That's how @haixuanTao should be able to map to wgpu. But another option is to implement new TypeOps that are GPU backed are replace the "standard" tract-core ones by these ops, leaving the others alone, then execute the model as usual. You can still call into_optimized actually, it will optimise whatever remains around the new operators you introduced.

**Reply by haixuanTao** [reply]
Just a quick note, that, you will not be able to replace everything that the current CPU code is doing only using GPU. You can also have a look at this issue to see the current limitation of WebGPU for Deep Learning: https://github.com/google/iree/issues/6499

**Reply by dsseng** [reply]
Will see the code and then contribute either integrating @ haixuanTao 's changes or doing own implementation.

**Reply by dsseng** [reply]
> Just a quick note, that, you will not be able to replace everything that the current CPU code is doing only using GPU. You can also have a look at this issue to see the current limitation of WebGPU for Deep Learning: [google/iree#6499](https://github.com/google/iree/issues/6499)

Which exactly? Will CPU-only parts be in the middle of the chain?

**Reply by dsseng** [reply]
Do you do such copies between RAM and VRAM in Wonnx?

**Reply by haixuanTao** [reply]
So, you might have to use a CPU for some ops for sure. In the current wonnx design, we do not implement this, but it's a clear limitation. You might not be able to compute `atanh` for example.

There are also memory and thread constraints on GPU that might force you to fallback on CPU as well. 

Yes, you can't read RAM from the GPU, you have to create a buffer between the two.

**Reply by dsseng** [reply]
`atanh(x)` is listed within WGSL reference for computation precision. We'll see what needs to be copied in between and check for any optimizations that might improve the perf (needs benches across platforms, btw, I posess Ryzen 5 4500U laptop with unified memory, so driver should just give the pointer to app I guess). If targeting mobiles I guess copying is not a problem at all because of UMA. Discrete desktop GPUs might require benchmark.

**Reply by haixuanTao** [reply]
Ok, so I didn't thought about integrated GPU and you actually can do a 0 copy apparently: https://www.reddit.com/r/rust/comments/s24uqt/unified_memory_on_m1_macs/?utm_medium=android_app&utm_source=share

I think they recently added `atanh`. My point was that some ops might not be easily implementable as there might not be an instruction within WGSL.

**Reply by dvc94ch** [reply]
advanced gpu features like subgroups aren't exposed in wgpu. but for now I guess getting it working is good enough, optimization can happen later. piet-gpu was so unsatisfied with wgpu that @raphlinus wrote his own gpu abstraction layer.

**Reply by dsseng** [reply]
Well if it goes fine we can then either use wgpu-hal or write direct Vulkan, Metal and DX12 backends. Definitely not for the start

**Comment by dsseng** [comment]
This returns proper model description, and then some backend methods should assemble this graph on GPU (allocate buffers, set BO properties to them (input and outputs, and maybe something that is computed on CPU should be accessible to it) and then execute on GPU. Am I right, or is there something else existing codebase can help me with and I skipped?

```
        tract_tensorflow::tensorflow()
        // load the model
        .model_for_path("mobilenet_v2_1.4_224_frozen.pb")?
        // specify input type and shape
        .with_input_fact(0, InferenceFact::dt_shape(f32::datum_type(), tvec!(1, 224, 224, 3)))?
        // optimize the model
        .into_typed()?
        .into_decluttered()?
```

**Reply by dsseng** [reply]
As far as I understood, `into_runnable` does create some sort of plan which is then CPU-bound. I didn't really understand how does the plan thing work, but it might be a universal mechanism I should not skip. @kali could you please elaborate to keep me from doing something useless.

**Reply by dsseng** [reply]
At least Plan needs to be forked to contain buffer pointers instead of buffers in RAM, but smth like order could be reused, but unsure whether that is worth it.

**Reply by kali** [reply]
The Plan is basically about finding the right order in which to call the operators, and some management of variables dependencies (we want to free intermediate tensors when we are done with them). This is established when you build up the Plan (so when into_runnable() is called). The plan is then an immutable shareable object.

In some cases, we also need a Session, to manage state (for stateful networks). But in the simple stateless case, the user can call Plan::run (it will create a throw-away session).

Now if you just substitute just some operators for more efficient ones before calling into_optimized().into_runnable(), they will implement the same EvalOp::eval method, so into_optimized() and the plan will keep working the same way on the rest of the network. If you the entire model replacement way, then once you've translated the TypedModel, what to do with it depends entirely on what you've implemented on your side. You may need to keep the TypedModel around or not butI don't think optimizing it or using the plan would make much sense at this stage.

**Reply by dsseng** [reply]
I do not plan to do EvalOp, since that will be less performant for discrete memory case. I want buffers to be allocated on gpu. So, I guess, I should take Plan, but not State.

**Reply by dsseng** [reply]
And SimpleState will be forked to GpuState in new crate `gpu`. Looks quite extensible, could be later used to play with MLIR or so. Thanks for getting those ideas to me!

**Comment by dsseng** [comment]
Am I going the right way? I'm still on the way of forking existing classes to accept GPUTensor instead of Tensor and use GPUOp instead of Op. The key difference is that GPUTensor should hold (doesn't now) pointer to VRAM, not actual data. I guess I'm stuck with doing something entirely wrong.
Diff of trials attached, it is not expected to be buildable:
[gpu-7.txt](https://github.com/sonos/tract/files/8597681/gpu-7.txt)

**Reply by dsseng** [reply]
I guess there should be a simpler way to make `Tensor` and `GPUTensor` generalized

**Reply by dsseng** [reply]
Hmm, looks like I will redo the whole idea of graphs etc. Just take decluttered model and pass backwards (what do I need to get output, what do I need to get 209 and 211 values etc), or use `SimplePlan` (which should contain necessary stuff already) counting memory objects on the way and making a mirror of graph which holds GPU data (buffers and shaders built to handle them).

**Reply by dsseng** [reply]
Btw is that a bug? When SimplePlan marks two ops as parallel, it keeps an empty element after:

```
        [
            80,
            91,
        ],
        [],
        [
            93,
        ],
```

**Reply by kali** [reply]
I don't think it is bug. There is no "marking ops as parallel" in SimplePlan, btw. I'm assuming this comes from the flush_list: it's the id of previous ops which results can be results once you're the same-index op.

or instance, if #0 is a Source, #1 a Source, #2 is Add(#0,#1), then the flush list should be [ [], [], [0,1] ] : when #2 is done, you can free outputs for #0 and #1.

**Reply by dsseng** [reply]
Okay. Is everything right with the idea of reading SimplePlan and then doing entirely GPU-specific stuff with own GPU ops?

**Reply by kali** [reply]
I think so. You may not even need the SimplePlan entirely. The eval_order function may be enough.

**Reply by dsseng** [reply]
Great. Now thinking about buffers with generic code. Need to allocate them for each step, pass over to next and drop those not necessary for further computations or output.

**Comment by dsseng** [comment]
Looks like #697 will now be the place to discuss the implementation as initial `GPUTensor` code is already there.
